akka-stream/src/main/scala/akka/stream/remote/scaladsl/SourceRef.scala

```diff
@@ -0,0 +1,215 @@
+ */
+package akka.stream.remote.scaladsl
+
+import akka.actor.{ ActorRef, Terminated }
+import akka.event.Logging
+import akka.stream._
+import akka.stream.actor.{ RequestStrategy, WatermarkRequestStrategy }
+import akka.stream.impl.FixedSizeBuffer
+import akka.stream.remote.StreamRefs
+import akka.stream.remote.impl.StreamRefsMaster
+import akka.stream.scaladsl.{ Sink, Source }
+import akka.stream.stage._
+import akka.util.OptionVal
+
+import scala.concurrent.{ Future, Promise }
+
+object SourceRef {
+
+  /**
+   * A local [[Sink]] which materializes a [[SourceRef]] which can be used by other streams (including remote ones),
+   * to consume data from this local stream, as if they were attached in the spot of the local Sink directly.
+   *
+   * Diagram: TODO a nice diagram
+   */
+  def sink[T](): Graph[SinkShape[T], Future[SourceRef[T]]] =
+    Sink.fromGraph(new SinkRef[T](OptionVal.None, materializeSourceRef = true))
+
+  // TODO Implement using TCP
+  // def bulkTransfer[T](): Graph[SinkShape[ByteString], SourceRef[ByteString]] = ???
+
+  implicit def convertRefToSource[T](ref: SourceRef[T]): Source[T, Future[SinkRef[T]]] =
+    Source.fromGraph(ref)
+}
+
+/**
+ * This stage can only handle a single "sender" (it does not merge values);
+ * The first that pushes is assumed the one we are to trust
+ */
+final class SourceRef[T](private[akka] val initialOriginRef: OptionVal[ActorRef]) extends GraphStageWithMaterializedValue[SourceShape[T], Future[SinkRef[T]]] {
+
+  val out: Outlet[T] = Outlet[T](s"${Logging.simpleName(getClass)}.out")
+  override def shape = SourceShape.of(out)
+
+  /**
+   * Convenience method for obtaining a [[Source]] from this [[SourceRef]] which is a [[Graph]].
+   *
+   * Please note that an implicit conversion is also provided in [[SourceRef]].
+   */
+  def source = Source.fromGraph(this)
+  /**
+   * Method used for obtaining a [[akka.stream.javadsl.Source]] from this [[SourceRef]] which is a [[Graph]].
+   */
+  def getSource = akka.stream.javadsl.Source.fromGraph(this)
+
+  override def createLogicAndMaterializedValue(inheritedAttributes: Attributes): (GraphStageLogic, Future[SinkRef[T]]) = {
+    val promise = Promise[SinkRef[T]]()
+
+    val logic = new TimerGraphStageLogic(shape) with StageLogging with OutHandler {
+      private[this] lazy val streamRefsMaster = StreamRefsMaster(ActorMaterializerHelper.downcast(materializer).system)
+      private[this] lazy val settings = streamRefsMaster.settings
+
+      override protected lazy val stageActorName = streamRefsMaster.nextSinkRefTargetSourceName()
+      private[this] var self: GraphStageLogic.StageActor = _
+      private[this] implicit def selfSender: ActorRef = self.ref
+
+      // demand management ---
+      private val highDemandWatermark = 16
+
+      private var expectingSeqNr: Long = 0L
+      private var localCumulativeDemand: Long = 0L // initialized in preStart with settings.initialDemand
```

Comment:
why not set it here instead of in preStart?

Reply:
Since it's from settings that I obtain from getting them from the extension / actor system, so I need to `ActorMaterializerHelper.downcast(materializer).system)`, which I can't before `preStart`, so sadly has to be initialized like that (or be a lazy val, like the settings are)